File: 计算机图像学/labs/lab1.py
import sys
import cv2
import numpy as np
import matplotlib as mpl
from matplotlib import pyplot as plt

def trans(img, k, b):
    img2 = np.zeros(img.shape, dtype=img.dtype)

    n, m = img.shape
    for i in range(n):
        for j in range(m):
            img2[i][j] = round(img[i][j] * k + b) # auto adjust data type
    return img2

def scale(img, x1, y1, x2, y2):
    img2 = np.zeros(img.shape, dtype=img.dtype)

    n, m = img.shape
    for i in range(n):
        for j in range(m):
            x = img[i][j]
            if (x < x1):
                img2[i][j] = 0+y1 * x // x2 # add 0(int)  to adjust data type
            elif x <= x2:
                img2[i][j] = 0+(y2 - y1) * (x - x1) // (x2 - x1) + y1
            else:
                img2[i][j] = 0+(255 - y2) * (x - x2) // (255 - x2) + y2
    return img2

def histogram(img, high=200, low=20):
    img2 = np.zeros(img.shape, dtype=img.dtype)

    n, m = img.shape
    dic = {}
    for i in range(n):
        for j in range(m):
            x = img[i][j]
            if x in dic:
                dic[x] += 1
            else:
                dic[x] = 1

    tot = n*m
    keys = sorted(i for i in dic if low <= i <= high)
    values = [dic[k]/tot for k in keys]
    return keys,values

def histogram_equalize(img=cv2.imread('images/pout.bmp',cv2.IMREAD_GRAYSCALE)):
    n, m = img.shape
    dic = {}
    for i in range(n):
        for j in range(m):
            x = img[i][j]
            if x in dic:
                dic[x] += 1
            else:
                dic[x] = 1

    acc = 0
    mapping={}
    tot = n*m
    for k in sorted(dic.keys()):  # sort, important
        acc+=dic[k]
        mapping[k]=round(acc/tot*255)

    equalized = {}
    for k in dic:
        newK = mapping[k]
        if newK in equalized:
            equalized[newK]+=dic[k]
        else:
            equalized[newK]=dic[k]
    img2 = np.zeros(img.shape, dtype=img.dtype)
    for i in range(n):
        for j in range(m):
            img2[i][j] = mapping[img[i][j]]
    return img2, equalized.keys(),[equalized[k]/tot for k in equalized]

def show(img, s='opencv'):
    print(s)
    cv2.namedWindow(s)
    cv2.resizeWindow(s, 1000, 1000)
    cv2.imshow(s, img)
    key = cv2.waitKey(5000)
    cv2.destroyAllWindows()


if __name__ == '__main__':
    path = sys.argv[1]
    img = cv2.imread(path, cv2.IMREAD_GRAYSCALE)

    k, b = 0.8, 10
    img2 = trans(img, k, b)

    x1, y1, x2, y2 = 5,5,200,150
    img3 = scale(img, x1, y1, x2, y2)

    high,low = 200,20
    keys,values = histogram(img,high,low)

    img4, keys2,values2 = histogram_equalize(img)

    cmap = mpl.cm.gray # mpl.cm.gray_r   'gray'

    plt.figure(figsize=(10, 10))

    plt.subplot(321), plt.imshow(img,cmap=cmap), plt.title('origin'),plt.xticks([]), plt.yticks([])
    plt.subplot(322), plt.imshow(img2,cmap=cmap), plt.title(f'tran k={k},b={b}'),plt.xticks([]), plt.yticks([])
    plt.subplot(323), plt.imshow(img3,cmap=cmap), plt.title(f'scaling  (x1,y1,x2,y2)=({(x1,y1,x2,y2)}'),plt.xticks([]), plt.yticks([])
    plt.subplot(324), plt.imshow(img4,cmap=cmap), plt.title(f'equalized'),plt.xticks([]), plt.yticks([])
    plt.subplot(325),plt.xlim([0,256]),  plt.bar(keys,values), plt.title(f'pdf')
    plt.subplot(326),plt.xlim([0,256]),  plt.bar(keys2,values2), plt.title(f'equalized pdf')

    plt.show()

File: 计算机图像学/labs/cpp/lab1.cpp
#include<opencv2/opencv.hpp>
#include<iostream>
#include<string>
#include<map>

using namespace std;
using namespace cv;

void trans(Mat & img, int k, int b)
{
    Mat img2 = Mat::zeros(img.size(), img.type());

    for (int i = 0; i < img.rows; ++i)
        for (int j = 0; j < img.cols; ++j) {
            int n = img.at<uchar>(i, j) * k + b;
            img2.at<uchar>(i, j) = n > 255 ? 255 : n;
        }

    namedWindow("transformed", 1);
    imshow("transformed", img2);
    waitKey();
}

void scale(Mat&img, uchar x1, uchar y1, uchar x2, uchar y2)
{
    Mat img2 = Mat::zeros(img.size(), img.type());

    for (int i = 0; i < img.rows; ++i)
        for (int j = 0; j < img.cols; ++j) {
            uchar x =  img.at<uchar>(i, j);

            if (x < x1) {
                img2.at<uchar>(i, j) = y1 / x1 * x;
            } else if (x1 <= x && x <= x2) {
                img2.at<uchar>(i, j) = (y2 - y1) / (x2 - x1) * (x - x1) + y1;
            } else {
                img2.at<uchar>(i, j) = (255 - y2) / (255 - x2) * (x - x2) + y2;
            }
        }

    namedWindow("scaleing", 1);
    imshow("scaling", img2);
    waitKey();
}

void histogram(Mat&img, uchar high, uchar low)
{
    int channels[] = {0};
    int nHistSize[] = {high - low + 1};
    float range[] = {low, high};
    const float* fHistRanges[] = {range};
    Mat hist;
    /*
     * images： 多幅图，只要深度相同，通道可以不同
     * channels： 通道数， 从0开始，
     * histSize: 直方图中每个维度级别数量, 即将 多少个灰度作为一个横坐标
     CV_EXPORTS void calcHist( const Mat* images,
     int nimages,
     const int* channels,
     InputArray mask,
     OutputArray hist,
     int dims,
     const int* histSize,
     const float** ranges,
     bool uniform = true,
     bool accumulate = false );
     */
    calcHist(&img, 1, channels, Mat(), hist, 1, nHistSize, fHistRanges, true);
    // canvas
    int nHistWidth = 800, nHistHeight = 600;
    int nBinWidth = (nHistWidth+nHistSize[0]-1) / nHistSize[0];
    Mat matHistImage(nHistHeight, nHistWidth, CV_8UC3, Scalar(255, 255, 255));
    //uniform
    normalize(hist, hist, 0.0, matHistImage.rows, NORM_MINMAX, -1, Mat());

    //draw histogram
    for (int i = 1; i < nHistSize[0]; ++i) {
        line(matHistImage, Point(nBinWidth * (i - 1), nHistHeight - cvRound(hist.at<float>(i - 1))), Point(nBinWidth * (8), nHistHeight - cvRound(hist.at<float>(i))), Scalar(255, 0, 0), 2, 8.0);
    }

    imshow("histogram", matHistImage);
    waitKey();
}


void histogram_enhance(Mat & img)
{
    cout << "before histogram enhancing..." << endl;
    namedWindow("before histogram enhancing...", 1);
    imshow("before", img);
    waitKey();
    int L1 = 255;
    long tot = img.rows * img.cols;
    //get cdf: cumulletive  distribute func
    map<uchar, long> cdf;

    for (int i = 0; i < img.rows; ++i)
        for (int j = 0; j < img.cols; ++j) {
            uchar p = img.at<uchar>(i, j);

            if (cdf.find(p) == cdf.end())
                cdf[p] = 1;
            else
                cdf[p] += 1;
        }

    int acc = 0;
    long min_cdf = tot;

    for (auto it = cdf.begin(); it != cdf.end(); ++it) {
        //map is sorted defaultly
        if (it->second < min_cdf)
            min_cdf = it->second;

        it->second += acc;
        acc = it->second;
    }

    double N1 = tot - min_cdf;
    //get gray mapping
    map<uchar, uchar> gray_map;

    for (auto it = cdf.begin(); it != cdf.end(); ++it) {
        gray_map[it->first] = (int)((it->second - min_cdf) * L1 / N1 + 0.5) ;
    }

    //substitute  and enhance
    Mat img2 = Mat::zeros(img.size(), img.type());

    for (int i = 0; i < img.rows; ++i)
        for (int j = 0; j < img.cols; ++j)
            img2.at<uchar>(i, j) = gray_map[ img.at<uchar>(i, j)];

    imshow("enhanced", img2);
    waitKey();
}

int main(int argc, char** argv)
{
    cout << "Input file path: ";
    string path;
    cin >> path;
    Mat img = imread(path, IMREAD_GRAYSCALE);

    cout << "Linear Transforming..." << endl;
    cout << "Input k,b: ";
    int k, b;
    cin >> k >> b;
    trans(img, k, b);

    cout << "Scaling..." << endl;
    cout << "Input x1,y1,x2,y2: ";
    int x1, x2, y1, y2;
    cin >> x1 >> y1 >> x2 >> y2;
    scale(img, x1, y1, x2, y2);

    cout << "Creating Histogram..." << endl;
    cout << "Input range low,high: ";
    int low, high;
    cin >> low, high;
    histogram(img, low, high);

    Mat mat = imread("images/pout.bmp", IMREAD_GRAYSCALE);
    histogram_enhance(mat);
    return 0;
}
